File: 76. Minimum Window Substring/main.py
import sys
from collections import defaultdict


class Solution:
    def minWindow(self, s: str, t: str) -> str:
        tMap = defaultdict(int)

        for c in t:
            tMap[c] += 1

        left = 0
        right = 0
        minLeft = 0
        minLen = sys.maxsize
        sLen = len(s)
        tLen = len(t)
        matched = 0
        sMap = defaultdict(int)

        for right in range(sLen):
            if s[right] not in tMap:
                continue

            sMap[s[right]] += 1
            if sMap[s[right]] <= tMap[s[right]]:
                matched += 1

            if matched == tLen:
                while s[left] not in tMap or sMap[s[left]] > tMap[s[left]]:
                    if s[left] in tMap:
                        sMap[s[left]] -= 1

                    left += 1

                size = right - left + 1
                if size < minLen:
                    minLen = size
                    minLeft = left

        if minLen == sys.maxsize:
            return ""

        return s[minLeft : minLeft + minLen]


class Solution:
    def minWindow(self, s: str, t: str) -> str:
        ARRAY_SIZE = 128
        count = [0] * ARRAY_SIZE

        for c in t:
            count[ord(c)] += 1

        left = 0
        right = 0
        minLeft = 0
        minLen = sys.maxsize
        matched = 0
        sLen = len(s)
        tLen = len(t)

        for right in range(sLen):
            if count[ord(s[right])] > 0:
                matched += 1

            count[ord(s[right])] -= 1

            while matched == tLen:
                size = right - left + 1
                if size < minLen:
                    minLen = size
                    minLeft = left

                count[ord(s[left])] += 1
                if count[ord(s[left])] > 0:
                    matched -= 1

                left += 1

        if minLen == sys.maxsize:
            return ""

        return s[minLeft : minLeft + minLen]


class Solution:
    def minWindow(self, s: str, t: str) -> str:
        ARRAY_SIZE = 128
        count = [0] * ARRAY_SIZE

        for c in t:
            count[ord(c)] += 1

        left = 0
        right = 0
        minLeft = 0
        minLen = sys.maxsize
        needed = len(t)
        sLen = len(s)

        while left < sLen:
            if right >= sLen and needed > 0:
                break

            if needed > 0:
                if count[ord(s[right])] > 0:
                    needed -= 1

                count[ord(s[right])] -= 1
                right += 1
            else:
                size = right - left
                if size < minLen:
                    minLen = size
                    minLeft = left

                count[ord(s[left])] += 1
                if count[ord(s[left])] > 0:
                    needed += 1

                left += 1

        if minLen == sys.maxsize:
            return ""

        return s[minLeft : minLeft + minLen]

File: 76. Minimum Window Substring/main.rs
struct Solution;

use std::collections::HashMap;

impl Solution {
    pub fn min_window(s: String, t: String) -> String {
        let s_vec = s.chars().collect::<Vec<_>>();
        let t_vec = t.chars().collect::<Vec<_>>();
        let mut t_map = HashMap::new();

        for &c in t_vec.iter() {
            *t_map.entry(c).or_insert(0) += 1;
        }

        let t_len = t.len();
        let mut left = 0;
        let mut matched = 0;
        let mut min_len = usize::MAX;
        let mut min_left = 0;
        let mut s_map = HashMap::new();

        for (right, &c) in s_vec.iter().enumerate() {
            if !t_map.contains_key(&c) {
                continue;
            }

            *s_map.entry(c).or_insert(0) += 1;

            if s_map[&c] <= t_map[&c] {
                matched += 1;
            }

            if matched == t_len {
                while !t_map.contains_key(&s_vec[left]) || s_map[&s_vec[left]] > t_map[&s_vec[left]]
                {
                    if t_map.contains_key(&s_vec[left]) {
                        *s_map.get_mut(&s_vec[left]).unwrap() -= 1;
                    }

                    left += 1;
                }

                let size = right - left + 1;

                if size < min_len {
                    min_len = size;
                    min_left = left;
                }
            }
        }

        if min_len == usize::MAX {
            return "".to_string();
        }

        s[min_left..min_left + min_len].into()
    }
}

impl Solution {
    pub fn min_window(s: String, t: String) -> String {
        let s_vec = s.chars().collect::<Vec<_>>();
        let t_vec = t.chars().collect::<Vec<_>>();
        const ARRAY_SIZE: usize = 128;
        let mut count = vec![0; ARRAY_SIZE];

        for &c in t_vec.iter() {
            count[c as usize] += 1;
        }

        let s_len = s_vec.len();
        let t_len = t_vec.len();
        let mut right = 0;
        let mut left = 0;
        let mut needed = t_len;
        let mut min_len = usize::MAX;
        let mut min_left = 0;

        while left < s_len {
            if right >= s_len && needed > 0 {
                break;
            }

            if needed > 0 {
                if count[s_vec[right] as usize] > 0 {
                    needed -= 1;
                }
                count[s_vec[right] as usize] -= 1;
                right += 1;
            } else {
                let size = right - left;
                if size < min_len {
                    min_len = size;
                    min_left = left;
                }

                count[s_vec[left] as usize] += 1;
                if count[s_vec[left] as usize] > 0 {
                    needed += 1;
                }
                left += 1;
            }
        }

        if min_len == usize::MAX {
            return "".to_string();
        }

        s[min_left..min_left + min_len].into()
    }
}

impl Solution {
    pub fn min_window(s: String, t: String) -> String {
        let s_bytes = s.as_bytes();
        let t_bytes = t.as_bytes();
        const ARRAY_SIZE: usize = 128;
        let mut count = vec![0; ARRAY_SIZE];

        for &c in t_bytes {
            count[c as usize] += 1;
        }

        let s_len = s_bytes.len();
        let t_len = t_bytes.len();
        let mut right = 0;
        let mut left = 0;
        let mut needed = t_len;
        let mut min_len = usize::MAX;
        let mut min_left = 0;

        while left < s_len {
            if right >= s_len && needed > 0 {
                break;
            }

            if needed > 0 {
                if count[s_bytes[right] as usize] > 0 {
                    needed -= 1;
                }
                count[s_bytes[right] as usize] -= 1;
                right += 1;
            } else {
                let size = right - left;
                if size < min_len {
                    min_len = size;
                    min_left = left;
                }

                count[s_bytes[left] as usize] += 1;
                if count[s_bytes[left] as usize] > 0 {
                    needed += 1;
                }
                left += 1;
            }
        }

        if min_len == usize::MAX {
            return "".to_string();
        }

        s[min_left..min_left + min_len].into()
    }
}


impl Solution {
    pub fn min_window(s: String, t: String) -> String {
        let s_bytes = s.as_bytes();
        let t_bytes = t.as_bytes();
        const ARRAY_SIZE: usize = 128;
        let mut count = vec![0; ARRAY_SIZE];

        for &c in t_bytes {
            count[c as usize] += 1;
        }

        let s_len = s_bytes.len();
        let t_len = t_bytes.len();
        let mut left = 0;
        let mut matched = 0;
        let mut min_len = usize::MAX;
        let mut min_left = 0;

        for right in 0..s_len {
            if count[s_bytes[right] as usize] > 0 {
                matched += 1;
            }

            count[s_bytes[right] as usize] -= 1;

            while matched == t_len {
                let size = right - left + 1;

                if size < min_len {
                    min_len = size;
                    min_left = left;
                }

                count[s_bytes[left] as usize] += 1;

                if count[s_bytes[left] as usize] > 0 {
                    matched -= 1;
                }
                left += 1;
            }
        }

        if min_len == usize::MAX {
            return "".to_string();
        }

        s[min_left..min_left + min_len].into()
    }
}
